### ISSUE TITLE
Multiple Decorators with different colours

### ISSUE THREAD

**Issue by LeonBre** [prolificinteractive/material-calendarview#617]
Hey there, 

I want to set the background of some dates to different color. For example I want to set the background of the 20.7.17 and of the 24.7.17 to red and the background of the 26. and 30. to green. 
I tried to do this with multiple decorators, but they only color the dates in one color.
Is there a possibility to do this ?asd

I used `addDecorators` and my decorator class looks like this:

    private final int mColor;
    private final HashSet<CalendarDay> dates;
    private Context mContext;

    EventDecorator(int color, Collection<CalendarDay> dates, Context context) {
        this.mColor = color;
        this.dates = new HashSet<>(dates);
        this.mContext = context;
    }

    @Override
    public boolean shouldDecorate(CalendarDay day) {
        return dates.contains(day);
    }

    @Override
    public void decorate(DayViewFacade view) {
        GradientDrawable drawable = (GradientDrawable) ResourcesCompat.getDrawable(mContext.getResources(), R.drawable.circle, null);
        if (drawable != null) {
            drawable.setColor(mColor);
            view.setBackgroundDrawable(drawable);

        }

    }

**Comment by Qwin** [comment]
Hi LeonBre,

I think I might be able to help you out, considering I was stuck myself and had to go through the whole calendar code. 

"A DayViewDecorator is an interface that has only two methods you need to implement, shouldDecorate(CalendarDay) and decorate(DayViewFacade). shouldDecorate() is called for every date in the calendar to determine if the decorator should be applied to that date. **decorate() is called only one time** to gather the customizations used for this decorator. This is so we can cache the decorations and efficiently apply them to many days." 

As per docs the decorator only is called ones and is cached after that, the color cannot be different within the same decorator. However creating multiple decorators with each a color should have worked, I did the same with my calendar and it works for me. (Unless I misunderstood you) 

Hope this helps.

**Comment by novas1r1** [comment]
Hey Qwin,

I added multiple Decorators to one day but at the end they are overlapping each other. I'm not able to set them next to each other. How did you solve that?

Have a nice day,
Nova

**Comment by LeonBre** [comment]
Hey @Qwin 
> However creating multiple decorators with each a color should have worked, I did the same with my calendar and it works for me.
I added 4 decorators with different colors to a material calendar view, but every date has the same color. Can you maybe give me your sample on which you achived this? This would help me a lot.

**Comment by Ruben2112** [comment]
@Qwin care to elaborate? I'm having the same issue but your solution doesn't seem to work for us.

**Comment by ka05** [comment]
@Qwin, I am also trying to achieve the same thing and cannot figure it out. 
Heres an example of what Im going for:

![screen shot 2018-07-03 at 9 53 26 am](https://user-images.githubusercontent.com/8964925/42223945-f99d01d0-7ea6-11e8-9c8c-1fed1a860261.png)

It looks like the `DotSpan` is overlapping because of the way the `void applyFacade(DayViewFacade facade)` method in the `DayView` class was written.

In the loop the span being set is always starting at 0 and going until the length of the label which is the day number of the month. Therefore the last one added will always be the one on top, effectively overwriting the others, rather than laying them out horizontally in a row.

```
String label = getLabel();
SpannableString formattedLabel = new SpannableString(getLabel());
for (DayViewFacade.Span span : spans) {
   formattedLabel.setSpan(span.span, 0, label.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
}
setText(formattedLabel);
```

This appears virtually unfixable based on the way that spans are added. Unless they are drawn in a separate view apart from the `DayView`, which is a `AppCompatCheckedTextView`, then they will always overlap.

Is there any other way to implement what I am trying to achieve?

**Comment by quentin41500** [comment]
@ka05 you are correct, the issue is related on how the spans are added. Im not a big fan of spans in general due to the limitation of customizing them, but they are more efficient for performance and less prone to weird behavior. If any of you find a way to add the dot spans without having them overlapping each other (and not breaking other spans), please let me know.

**Comment by vanphuc0503** [comment]
@ka05 How to draw a calendar like you, can you send me an example, thanks guy
